### ISSUE TITLE
"Given" step randomly fails to run

### ISSUE THREAD

**Issue by jcholder-thoughtworks** [bbqsrc/cucumber-rust#44]
Howdy! I have an issue where a limited set of my `Given` steps randomly fail to run.

I verified this by tossing a `panic!` into one such step then watching the step fail on some runs and pass on others despite no other code changes between those runs.

The step in question:

```rust
given regex r"transform ← translation\((.*), (.*), (.*)\)" |world, matches, _step| {
    let t1: f32 = matches[1].parse().unwrap();
    let t2: f32 = matches[2].parse().unwrap();
    let t3: f32 = matches[3].parse().unwrap();

    world.transform = translation(t1, t2, t3).matrix;
    panic!("Called!");
};
```

`world` is the `MyWorld` from the example code, just expanded.

Any idea what could make things fail non-deterministically? I didn't notice any randomness or caching in the code here that might explain it. Thanks! ([The example setup step in context](https://github.com/jcholder-thoughtworks/ray_tracer_challenge/blob/master/tests/cucumber.rs#L236.))

**Comment by bbqsrc** [comment]
Do you have a repo I can use to reproduce this?

**Comment by jcholder-thoughtworks** [comment]
@bbqsrc Here you go: https://github.com/jcholder-thoughtworks/ray_tracer_challenge.

**Comment by bbqsrc** [comment]
Thanks! Can confirm the bug is happening, super weird. Will look into this, give me a few days.

**Comment by jcholder-thoughtworks** [comment]
Much appreciated!

**Comment by bbqsrc** [comment]
Turns it's not a bug per se, just how you've implemented your regular expressions.

The hash map is keyed by the regular expressions, which is ordered by hash and will differ between runs. Correctly written regularly expressions will make this order not matter.

`Given transform ← scaling(-1, 1, 1)` is picked up by either `m ← scaling\((.*), (.*), (.*)\))` or `transform ← scaling\((.*), (.*), (.*)\))`. Best fix is to prefix your regexes with `^` and end with `$` to ensure full matches only. 🙂

**Comment by bbqsrc** [comment]
Published a fix to the non-deterministic feeling by changing from HashMap to BTreeMap. You should still fix your regular expressions though ;D

Published as v0.6.6

**Comment by jcholder-thoughtworks** [comment]
@bbqsrc Thanks! The Perl programmer in me is embarrassed by that oversight.
